PFW-1504 fix conflicting UI issues

gudnimg opened the pull request:
Proposal to fix some of the issues with the initial implementation it is safer to use the status line code to print the message so there aren't any conflicts in the LCD cursor position.

Allow inserting a byte into any position in the LCD status message

Also, add a variable to control from which index in the array should the message start printing. This is very useful for progress bars and messages which continually update. I think we can save some memory by applying this to Mesh Bed Leveling later.

Change in memory:
Flash: +208 bytes
SRAM: +1 byte

gudnimg committed:
PFW-1504 fix conflicting UI issues

Proposal to fix some of the issues with the initial implementation
it is safer to use the status line code to print the message so
there aren't any conflicts in the LCD cursor position.

Allow inserting a byte into any position in the LCD status message

Also, add a variable to control from which index in the array
should the message start printing. This is very useful for progress
bars and messages which continually update. I think we can save some
memory by applying this to Mesh Bed Leveling later.

Change in memory:
Flash: +106 bytes
SRAM: +1 byte

gudnimg committed:
PFW-1504 Cleanup

Only set cursor with lcd_status_message_idx where the variable
is used.

No change in memory

gudnimg commented:
Got everything working finally when printing and running toolchanges. Synced with MK3 branch (will hopefully re-test tomorrow)

gudnimg committed:
Previously we called cleared the LCD, no need to change that

Change in memory:
Flash: -6 bytes
SRAM: 0 bytes

gudnimg commented:
Oof, after a rebase to sync with MK3 branch. The flash consumption increased from 94B to 114B. 🤔

gudnimg commented:
Converting to draft. The solution for the 4 second delay in https://github.com/prusa3d/Prusa-Firmware/pull/4098 is not very nice and not even in my initial PFW-1504 PR (though it 'works'). It tempting to re-write the status line code but I probably don't have time to do it :) + I would 100% be creating more bugs than solving existing ones.

I also don't like how much memory this PR uses.

Need to think more about this.

gudnimg commented:
Adding to this PR discussion comments from @wavexx in https://github.com/prusa3d/Prusa-Firmware/pull/4098

> Reading through commit https://github.com/prusa3d/Prusa-Firmware/commit/bbfc4b9f7853525d36a74e09ffc14366bcd9edbc it seems that we touch `lcd_status_message_timeout` with the intention to avoid the display to be updated so that we can keep the previous changes for the filament status on-screen.

> 
> `lcd_status_message_timeout` is used to prevent the message from being changed depending on priority, but it still needs to be displayed in this function. This is what is causing the screen to become unresponsive then the thermal timeout is happening. We can't use the timer like this, and we can't combine it for this purpose.

> 
> `INFO` needs to have higher priority, and can't be expired by something like the filament sensor.

> 
> We could introduce a new message priority (something like `LCD_STATUS_NOCHANGE`) inbetween `INFO` and `NONE`, and handle that with a different timeout. The priority needs to be handled in `lcd_message_check` (which currently only handles `INFO`, since other priorities do not expire) and use it to see if you can actually set/reset the message priority and then expire it properly.
> 
> I am referring to the progress drawn by `TryLoadUnloadProgressbar()`, just to clarify.

gudnimg committed:
PFW-1504 Don't clear Status line on Deinit

Status line code should decide what to render next

Change in memory:
Flash: -12 bytes
SRAM: 0 bytes

gudnimg commented:
Feel free to review this PR for RC1/RC2. It works so far, only thing left to do is find a better solution for the time delay. I won't be able to look into it until next week.

gudnimg commented:
Should we move this to RC2?

leptun commented:
If it's not ready till tomorrow, then yes. I'd like to avoid blocking the RC1 testing task on Monday unless something is broken and needs fixing asap.

gudnimg commented:
The issue addressed in this PR is not a blocker IMO. It's only UI related. The LCD visualization code needs to be changed anyway due to changes on the MK4 side so I think it's best to wait until then.

gudnimg commented:
Dropped one commit. Also synced branch to latest MK3.

gudnimg committed:
Echo the result onto serial

Example:
MMU2:1111111111111110011
1 means filament present (solid block)
0 means otherwise (dash)

Change in memory:
Flash: +94 bytes
SRAM: 0 bytes

gudnimg commented:
Rebased to sync the branch. Resolved one conflict.

gudnimg commented:
Rebased to sync with MK3 branch (48 commits)

3d-gussner reviewed (approved):
Looks good to me
